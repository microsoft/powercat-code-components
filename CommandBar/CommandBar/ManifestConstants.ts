export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Enabled = 'ItemEnabled',
    IconName = 'ItemIconName',
    IconColor = 'ItemIconColor',
    Overflow = 'ItemOverflow',
    FarItem = 'ItemFarItem',
    ParentKey = 'ItemParentKey',
    IconOnly = 'ItemIconOnly',
    Visible = 'ItemVisible',
    Checked = 'ItemChecked',
    Split = 'ItemSplit',
    ItemHeader = 'ItemHeader',
    ItemTopDivider = 'ItemTopDivider',
    ItemDivider = 'ItemDivider',
    ItemStyles = 'ItemStyles',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}
