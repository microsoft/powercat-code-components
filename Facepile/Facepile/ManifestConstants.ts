export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemPersonaName',
    Key = 'ItemPersonaKey',
    ImageUrl = 'ItemPersonaImageUrl',
    Image = 'ItemPersonaImage',
    ImageInfo = 'ItemPersonaImageInfo',
    Role = 'ItemPersonaRole',
    ImageType = 'ItemImageType',
    Presence = 'ItemPersonaPresence',
    Clickable = 'ItemPersonaClickable',
    IsImage = 'IsImage',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
    SelectedKey = 'SelectedKey',
}

export const enum OutputEvents {
    PersonaEvent = 'PersonaEvent',
    OverFlowButtonEvent = 'OverFlowButtonEvent',
    AddButtonEvent = 'AddButtonEvent',
}
