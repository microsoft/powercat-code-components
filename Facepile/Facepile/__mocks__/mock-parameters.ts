/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockEnumProperty, MockStringProperty, MockTwoOptionsProperty, MockWholeNumberProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        AccessibilityLabel: new MockStringProperty(),
        InputEvent: new MockStringProperty(),
        Theme: new MockStringProperty(),
        items: new MockDataSet([]),
        MaxDisplayablePersonas: new MockWholeNumberProperty(),
        ImageShouldFadeIn: new MockTwoOptionsProperty(),
        PersonaSize: new MockEnumProperty(),
        OverflowButtonType: new MockEnumProperty(),
        ShowAddButton: new MockTwoOptionsProperty(),
        OverflowButtonAriaLabel: new MockStringProperty(),
        AddbuttonAriaLabel: new MockStringProperty(),
    };
}
