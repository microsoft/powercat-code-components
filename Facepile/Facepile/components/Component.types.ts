import { PersonaSize, IFacepilePersona, OverflowButtonType } from '@fluentui/react';

export interface IFacepileprops {
    width?: number;
    height?: number;
    text?: string;
    textColor?: string;
    imagesFadeIn?: boolean;
    displayedPersonas?: number;
    personaSize: PersonaSize;
    items: ICustomFacepile[];
    disabled?: boolean;
    tabIndex?: number;
    ariaLabel?: string;
    themeJSON?: string;
    setFocus?: string;
    overflowButtonType?: OverflowButtonType;
    onSelected: (eventName: string, item?: IFacepilePersona) => void;
    showAddButton: boolean;
    overflowButtonAriaLabel: string;
    addbuttonAriaLabel: string;
}

export interface ICustomFacepile {
    id: string;
    key: string;
    personaName: string;
    imageUrl: string;
    presence: string;
    imageInitials?: string;
    initialsColor?: number;
    clickable: boolean;
    onClick?: (ev: unknown, persona?: IFacepilePersona) => void;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data?: any;
}
