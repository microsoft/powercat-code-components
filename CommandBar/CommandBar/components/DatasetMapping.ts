import { CanvasCommandItem } from './Component.types';
import {
    ContextualMenuItemType,
    getColorFromString,
    getShade,
    IButtonStyles,
    ICommandBarItemProps,
    IContextualMenuItem,
    IIconProps,
    Shade,
} from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';

export function getMenuItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): CanvasCommandItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = (record.getFormattedValue(ItemColumns.Key) as string) ?? '';
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getFormattedValue(ItemColumns.DisplayName),
            enabled: defaultIfNull(record.getValue(ItemColumns.Enabled) as boolean, true),
            iconName: record.getFormattedValue(ItemColumns.IconName),
            iconColor: record.getFormattedValue(ItemColumns.IconColor),
            farItem: undefinedIfNull(record.getValue(ItemColumns.FarItem) as boolean),
            overflow: undefinedIfNull(record.getValue(ItemColumns.Overflow) as boolean),
            parentItemKey: record.getFormattedValue(ItemColumns.ParentKey),
            iconOnly: undefinedIfNull(record.getValue(ItemColumns.IconOnly) as boolean),
            checked: undefinedIfNull(record.getValue(ItemColumns.Checked) as boolean),
            visible: defaultIfNull(record.getValue(ItemColumns.Visible) as boolean, true),
            split: undefinedIfNull(record.getValue(ItemColumns.Split) as boolean),
            isHeader: (record.getValue(ItemColumns.ItemHeader) as boolean) ?? undefined,
            topDivider: (record.getValue(ItemColumns.ItemTopDivider) as boolean) ?? undefined,
            divider: (record.getValue(ItemColumns.ItemDivider) as boolean) ?? undefined,
            stylesJSON: (record.getValue(ItemColumns.ItemStyles) as string) ?? undefined,
            data: record,
        } as CanvasCommandItem;
    });
}

function getDummyAction(key: string): CanvasCommandItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as CanvasCommandItem;
}

function getButtonStyles(stylesJSON?: string): IButtonStyles {
    try {
        return (stylesJSON ? JSON.parse(stylesJSON) : {}) as IButtonStyles;
    } catch (ex) {
        console.error('Cannot parse command bar item style', ex);
        return {} as IButtonStyles;
    }
}

export function getCommandsWithChildren(
    items: CanvasCommandItem[],
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): ICommandBarItemProps[] {
    return items
        .filter((i) => !i.parentItemKey && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
}

export function getCommandBarItemProps(
    items: CanvasCommandItem[],
    item: CanvasCommandItem,
    disabled: boolean,
    onClick: (ev?: React.MouseEvent<HTMLButtonElement>, item?: IContextualMenuItem | undefined) => boolean,
): ICommandBarItemProps {
    const subItems = items
        .filter((i) => i.parentItemKey === item.key && i !== item && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });

    const buttonStyles = getButtonStyles(item.stylesJSON);

    const iconProps =
        item.iconName &&
        ({
            iconName: item.iconName,
        } as IIconProps);

    // If custom icon color, find the shade slots for the hover/pressed state
    if (iconProps && item.iconColor) {
        const iconColor = item.iconColor;
        iconProps.styles = { root: { color: item.iconColor + ' !important' } };

        // Attempt to set a darker shade for hover/pressed colors
        const primaryColor = getColorFromString(iconColor);
        if (primaryColor) {
            const iconColorDarker = getShade(primaryColor, Shade.Shade8);
            const secondaryColorStyle = { color: iconColorDarker?.str + ' !important' };
            buttonStyles.iconHovered = secondaryColorStyle;
            buttonStyles.iconPressed = secondaryColorStyle;
            buttonStyles.iconExpanded = secondaryColorStyle;
            buttonStyles.iconChecked = secondaryColorStyle;
            buttonStyles.iconExpandedHovered = secondaryColorStyle;
        }
    }

    const props = {
        key: item.key,
        text: item.name,
        disabled: item.enabled === false || disabled,
        onClick: onClick,
        iconOnly: item.iconOnly === true,
        iconProps: iconProps,
        split: item.split,
        canCheck: item.checked !== undefined,
        checked: item.checked,
        data: item,
        buttonStyles: buttonStyles,
        subMenuProps:
            subItems.length > 0
                ? {
                      items: subItems,
                  }
                : undefined,
    } as ICommandBarItemProps;

    if (item.isHeader === true && subItems.length > 0) {
        // Semantic Section
        props.itemType = ContextualMenuItemType.Section;
        props.sectionProps = {
            title: item.name,
            topDivider: item.topDivider,
            bottomDivider: item.divider,
            items: subItems,
        };
    } else if (item.isHeader === true) {
        // Simple Section Break
        props.itemType = ContextualMenuItemType.Header;
    } else if (item.divider === true) {
        // Divider Break
        props.itemType = ContextualMenuItemType.Divider;
    } else {
        props.itemType = ContextualMenuItemType.Normal;
        props.subMenuProps =
            subItems.length > 0
                ? {
                      items: subItems,
                  }
                : undefined;
    }
    return props;
}

function undefinedIfNull<T>(value: T | null) {
    return defaultIfNull(value, undefined);
}
function defaultIfNull<T>(value: T | null, defaultValue: T) {
    return value === null ? defaultValue : value;
}
