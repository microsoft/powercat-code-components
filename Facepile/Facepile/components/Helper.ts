import { ItemColumns } from '../ManifestConstants';
import { PersonaSize, OverflowButtonType, PersonaPresence } from '@fluentui/react';

export function getUrlfromImage(record: ComponentFramework.PropertyHelper.DataSetApi.EntityRecord): string {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const imageData = undefinedIfNullish(record.getValue(ItemColumns.Image) as any) ?? '';
    return imageData ? `data:image/jpeg;base64, ${imageData.fileContent}` : imageData;
}

export function getImageUrl(record: ComponentFramework.PropertyHelper.DataSetApi.EntityRecord): string {
    return (record.getValue(ItemColumns.ImageInfo) as string) ?? '';
}

export function undefinedIfNullish<T>(value: T): T | undefined {
    return defaultIfNullish(value, undefined);
}
function defaultIfNullish<T>(value: T, defaultValue: T) {
    return (value as T) ? value : defaultValue;
}

export function getPersonaPresence(record: ComponentFramework.PropertyHelper.DataSetApi.EntityRecord): PersonaPresence {
    const personaPresence = undefinedIfNullish(record.getValue(ItemColumns.Presence));
    if (personaPresence) {
        switch ((personaPresence as string).toLowerCase()) {
            case 'away':
                return PersonaPresence.away;
            case 'blocked':
                return PersonaPresence.blocked;
            case 'busy':
                return PersonaPresence.busy;
            case 'dnd':
                return PersonaPresence.dnd;
            case 'none':
                return PersonaPresence.none;
            case 'offline':
                return PersonaPresence.offline;
            case 'online':
                return PersonaPresence.online;
            default:
                return PersonaPresence.none;
        }
    } else {
        return PersonaPresence.none;
    }
}

export function getPersonaSize(sizeSelected: string): PersonaSize {
    switch (sizeSelected.toLowerCase()) {
        case 'size8':
            return PersonaSize.size8;
        case 'size24':
            return PersonaSize.size24;
        case 'size32':
            return PersonaSize.size32;
        case 'size40':
            return PersonaSize.size40;
        case 'size48':
            return PersonaSize.size48;
        case 'size56':
            return PersonaSize.size56;
        default:
            return PersonaSize.size32;
    }
}

export function getOverflowButtonType(btnTypeSelected: string): OverflowButtonType {
    switch (btnTypeSelected.toLowerCase()) {
        case 'none':
            return OverflowButtonType.none;
        case 'descriptive':
            return OverflowButtonType.descriptive;
        case 'downarrow':
            return OverflowButtonType.downArrow;
        case 'more':
            return OverflowButtonType.more;
        default:
            return OverflowButtonType.descriptive;
    }
}
