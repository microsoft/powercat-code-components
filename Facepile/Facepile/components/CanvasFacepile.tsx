import * as React from 'react';
import { Facepile as CustomFacepile } from '../fluenui-fork/Facepile/Facepile';
import { IFacepilePersona, ThemeProvider, createTheme, IPartialTheme } from '@fluentui/react';
import { IFacepileprops } from './Component.types';
import { getFacepilePersonas } from './DatasetMapping';
import { getPersonaPresence } from './Helper';
import { useAsync } from '@fluentui/react-hooks';
import { OutputEvents } from '../ManifestConstants';

export const CanvasFacepile = React.memo((props: IFacepileprops) => {
    const {
        width,
        height,
        imagesFadeIn,
        displayedPersonas,
        onSelected,
        personaSize,
        setFocus,
        themeJSON,
        ariaLabel,
        items,
        overflowButtonType,
        showAddButton,
        overflowButtonAriaLabel,
        addbuttonAriaLabel,
        tabIndex,
    } = props;

    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

    const getPersonaProps = React.useCallback(
        (persona: IFacepilePersona) => ({
            imageShouldFadeIn: imagesFadeIn,
            ...(typeof persona.data !== 'undefined' && { presence: getPersonaPresence(persona.data) }),
        }),
        [imagesFadeIn],
    );

    const overflowButtonProps = {
        ariaLabel: overflowButtonAriaLabel,
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onClick: (ev: React.MouseEvent<HTMLButtonElement>) => onSelected(OutputEvents.OverFlowButtonEvent),
    };

    const addButtonProps = {
        ariaLabel: addbuttonAriaLabel,
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        onClick: (ev: React.MouseEvent<HTMLButtonElement>) => onSelected(OutputEvents.AddButtonEvent),
    };

    // Provide an explicit size so that the ResizeGroup measurements are correct
    const rootStyle = React.useMemo(() => {
        return {
            display: 'block',
            position: 'absolute',
            width: width,
            height: height,
        } as React.CSSProperties;
    }, [width, height]);

    const onClick = React.useCallback(
        (ev?: unknown, item?: IFacepilePersona) => {
            const selectedItem = item && items.find((i: IFacepilePersona) => i.id === item?.id);
            if (selectedItem) onSelected(OutputEvents.PersonaEvent, selectedItem);
        },
        [items, onSelected],
    );

    const rootRef = React.useRef<HTMLDivElement>(null);
    const async = useAsync();
    React.useEffect(() => {
        if (setFocus && setFocus !== '' && rootRef) {
            async.requestAnimationFrame(() => {
                // We can't call focus() on the imperative componentRef because of a bug in ResizeGroup
                // that causes the componentRef.current to be nulled
                // See https://github.com/microsoft/fluentui/issues/22844
                const buttons = (rootRef.current as HTMLElement).getElementsByTagName('button');
                if (buttons && buttons.length > 0) {
                    buttons[0].focus();
                }
            });
        }
    }, [setFocus, rootRef, async]);

    const facepilePersonas: IFacepilePersona[] = getFacepilePersonas(items, onClick);

    return (
        <ThemeProvider theme={theme} style={rootStyle}>
            <CustomFacepile
                personaSize={personaSize}
                personas={facepilePersonas}
                maxDisplayablePersonas={displayedPersonas}
                overflowButtonType={overflowButtonType}
                overflowButtonProps={overflowButtonProps}
                getPersonaProps={getPersonaProps}
                ariaLabel={ariaLabel}
                showAddButton={showAddButton}
                addButtonProps={addButtonProps}
                tabIndex={tabIndex}
            />
        </ThemeProvider>
    );
});
CanvasFacepile.displayName = 'CanvasFacepile';
