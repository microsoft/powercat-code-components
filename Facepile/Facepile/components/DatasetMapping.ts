import { IFacepilePersona } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';
import { ICustomFacepile } from './Component.types';
import { undefinedIfNullish, getImageUrl, getUrlfromImage } from './Helper';

export function getitemFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomFacepile[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return getDummyAction();
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;

        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            personaName: record.getFormattedValue(ItemColumns.DisplayName),
            imageUrl: undefinedIfNullish(record.getValue(ItemColumns.IsImage) as boolean)
                ? getUrlfromImage(record)
                : getImageUrl(record),
            data: record,
            presence: record.getValue(ItemColumns.Presence) as string,
            clickable: (record.getValue(ItemColumns.Clickable) as boolean) ?? false,
        } as ICustomFacepile;
    });
}

export function getFacepilePersonas(
    items: ICustomFacepile[],
    onClick: (ev?: unknown, item?: IFacepilePersona) => void,
): ICustomFacepile[] {
    return items.map((item: ICustomFacepile) => ({
        id: item.id,
        key: item.key,
        personaName: item.personaName,
        presence: item.presence,
        imageUrl: item.imageUrl,
        ...(item.clickable && { onClick: onClick }),
    })) as ICustomFacepile[];
}

function getDummyAction(): ICustomFacepile[] {
    return [
        {
            id: '1',
            key: '1',
            personaName: 'Megan Bowen',
            imageUrl:
                'https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/persona-female.png',
            imageInitials: 'Item ' + 1,
            initialsColor: 1,
            presence: 'none',
        },
        {
            id: '2',
            key: '2',
            personaName: 'Diego Siciliano',
            imageUrl:
                'https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/persona-male.png',
            imageInitials: 'Item ' + 1,
            initialsColor: 1,
            presence: 'none',
        },
    ] as ICustomFacepile[];
}
