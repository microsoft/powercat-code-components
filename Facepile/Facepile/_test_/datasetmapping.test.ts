import { getFacepilePersonas, getitemFromDataset } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.DisplayName]: 'Diego Siciliani',
                [ItemColumns.ImageUrl]:
                    'https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/persona-male.png',
                [ItemColumns.Clickable]: true,
                [ItemColumns.Presence]: 'away',
            }),
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item2',
                [ItemColumns.DisplayName]: 'Megan Bowen',
                [ItemColumns.ImageUrl]:
                    'https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/persona-female.png',
                [ItemColumns.Clickable]: true,
                [ItemColumns.Presence]: 'none',
            }),
        ];

        const onClickEvent = jest.fn();
        const actions = getitemFromDataset(new MockDataSet(items));
        const props = getFacepilePersonas(actions, onClickEvent);
        expect(props).toMatchSnapshot();
    });
});
