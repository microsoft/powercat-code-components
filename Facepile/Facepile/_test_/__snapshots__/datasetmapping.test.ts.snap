// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "id": "1",
    "imageUrl": "",
    "key": "item2",
    "onClick": [MockFunction],
    "personaName": "Megan Bowen",
    "presence": "none",
  },
  Object {
    "id": "1",
    "imageUrl": "",
    "key": "item2_2",
    "onClick": [MockFunction],
    "personaName": "Megan Bowen",
    "presence": "none",
  },
]
`;
